File: LeetCode/1734. Decode XORed Permutation/solution.cpp
/**
 * 63 / 63 test cases passed.
 * Runtime: 180 ms
 * Memory Usage: 99.9 MB 
 */
class Solution {
public:
    vector<int> decode(vector<int>& encoded) {
        int first = 0;
        for (int i = 0; i <= encoded.size(); ++ i) {
            first ^= i + 1;
            if (i & 1) first ^= encoded[i];
        }
        vector<int> ans = { first };
        for (auto& num: encoded) {
            ans.push_back( ans.back() ^ num );
        }
        return ans;
    }
};

File: LeetCode/1734. Decode XORed Permutation/solution.py
"""
63 / 63 test cases passed.
Runtime: 168 ms
Memory Usage: 30.8 MB
"""
class Solution:
    def decode(self, encoded: List[int]) -> List[int]:
        first = reduce(xor, encoded[1::2] + list(range(1, len(encoded) + 2)))
        return list(accumulate(encoded, xor, initial=first))

